### PR TITLE
reset events before apply runtime upgrade

### LINKED ISSUE #10603: Events from on_runtime_upgrade are unavailable
`on_runtime_upgrade` hook is executed before `initialize` and which initialize fields like block number as well as kill all the events from previous block. This means events from `on_runtime_upgrade` are also removed by `initialize`.

This makes debugging harder as well as make events incomplete. It will not be possible to observe changes by only look at events.

I can see reason why it is executed before `initialize` but most of the time, the reason doesn't apply and that means code from `on_runtime_upgrade` will have no event, read old block number, read bad execution phase, etc.

### PR DESCRIPTION
**Pull request by xlc**
closes #10603

polkadot companion: https://github.com/paritytech/polkadot/pull/4696
cumulus companion: https://github.com/paritytech/cumulus/pull/894

### CODE REVIEW

File: frame/system/src/lib.rs [hunk 1]
```diff
@@ -1446,7 +1417,6 @@ impl<T: Config> Pallet<T> {
 
 	/// Reset events. Can be used as an alternative to
 	/// `initialize` for tests that don't need to bother with the other environment entries.
```
Comment (bkchr) [inline frame/system/src/lib.rs:1419]: Docs need to be updated.

File: frame/executive/src/lib.rs [hunk 2]
```diff
@@ -294,6 +294,11 @@ where
 		parent_hash: &System::Hash,
 		digest: &Digest,
 	) {
+		// Reset events before apply runtime upgrade hook.
+		// This is required to preserve events from runtime upgrade hook.
+		// This means the format of all the event related storages must always be compatible.
+		<frame_system::Pallet<System>>::reset_events();
```
Comment (bkchr) [inline frame/executive/src/lib.rs:300]: Please add a test to ensure this behavior. There should already be some test for runtime upgrade below, just send some event and check it is present.

**Review by bkchr — COMMENTED** [review]
Looks good in general, just some nitpicks.

**Comment by bkchr** [comment]
It also doesn't compile.

File: frame/system/src/lib.rs [hunk 3]
```diff
@@ -1444,9 +1411,8 @@ impl<T: Config> Pallet<T> {
 		AllExtrinsicsLen::<T>::put(len as u32);
 	}
 
-	/// Reset events. Can be used as an alternative to
-	/// `initialize` for tests that don't need to bother with the other environment entries.
-	#[cfg(any(feature = "std", feature = "runtime-benchmarks", test))]
+	/// Reset events. This needs to be used in addition with `initialize` for each new block
+	/// to clear events from previous block.
```
Comment (bkchr) [inline frame/system/src/lib.rs:1415]:
```suggestion
	/// Reset events.
	///
	/// This needs to be used in prior calling [`initialize`](Self::initialize) for each new block
	/// to clear events from previous block.
```

**Review by bkchr — APPROVED** [review]

**Review by shawntabrizi — APPROVED** [review]

**Comment by bkchr** [comment]
@xlc we need a companion for this.

**Comment by bkchr** [comment]
bot merge

**Pull request merged**